cli/popper/runner_slurm.py

```diff
@@ -0,0 +1,155 @@
+import os
+from subprocess import PIPE, Popen, STDOUT, SubprocessError
+
+import docker
+
+from popper import utils as pu
+from popper import scm
+from popper.cli import log as log
+from popper.runner import StepRunner as StepRunner
+from popper.runner_host import DockerRunner as HostDockerRunner
+
+
+class SlurmRunner(StepRunner):
+
+    spawned_processes = []
+
+    def __init__(self, config):
+        super(SlurmRunner, self).__init__(config)
+
+    def __exit__(self, exc_type, exc, traceback):
+        SlurmRunner.spawned_processes = []
+
+    def exec_srun_cmd(self, cmd):
+        step_env = SlurmRunner.prepare_environment(self.step, os.environ)
+        try:
+            cmd.insert(0, 'srun')
+            with Popen(cmd, stdout=PIPE, stderr=STDOUT,
+                       universal_newlines=True, preexec_fn=os.setsid,
+                       env=step_env, cwd=self.config.workspace_dir) as p:
+                SlurmRunner.spawned_processes.append(p)
+
+                log.debug('Reading process output')
+
+                for line in iter(p.stdout.readline, ''):
+                    line_decoded = pu.decode(line)
+                    log.step_info(line_decoded[:-1])
+
+                p.wait()
+                ecode = p.poll()
+
+            log.debug(f'Code returned by process: {ecode}')
+
+        except SubprocessError as ex:
+            ecode = ex.returncode
+            log.step_info(f"Command '{cmd[0]}' failed with: {ex}")
+        except Exception as ex:
+            ecode = 1
+            log.step_info(f"Command raised non-SubprocessError error: {ex}")
+
+        return ecode
+
+    def stop_srun_cmd(self):
+        for p in SlurmRunner.spawned_processes:
+            log.info(f'Stopping proces {p.pid}')
+            p.kill()
+
+
+class HostRunner(SlurmRunner):
```

Comment:
we can remove this and instead add at the top:

```
from popper.runner_host import HostRunner
```

since the `HostRunner` would be the same as the one  in `runner_host.py`

Reply:
Actually, the `runner_slurm.HostRunner` is quite different than in`runner_host.HostRunner`. It just prepares the command and uses the methods defined in `runner_slurm.SlurmRunner` to execute the command.

Reply:
I see. And how do you envision this being used from the point of view a workflow? Were you thinking the `runner_slurm.HostRunner` would be used for `uses: "sh"` steps? For example, when this runs:

```yaml
- uses: sh
  runs: ls
```

the semantics is that the command will run directly on the machine where `popper run` is being executed, not via `srun`.

Reply:
Done